File: algorithms/graph/dfs.cpp
#include<bits/stdc++.h>
using namespace std;

void dfs(vector<vector<int>> graph){
  stack<int> s;
  bool visited[graph.size()] = {false};
  s.push(0);
  visited[0] = true;
  while(!s.empty()){
    int curr = s.top(); s.pop();
    cout << curr << " ";
    for(int i=0; i<graph[curr].size(); i++){
      if(visited[graph[curr][i]]) continue;
      visited[graph[curr][i]] = true;
      s.push(graph[curr][i]);
    }
  }
}

void dfs_rec(vector<vector<int>> graph, bool visited[], int root){
  if(visited[root]) return;
  visited[root] = true;
  for(int i=0; i<graph[root].size(); i++) dfs_rec(graph, visited, graph[root][i]);
  cout << root <<" ";
}
int main(){
  vector<vector<int>> graph = {{1,3},{0,2},{1,3,4},{0,2},{2}};
  dfs(graph);
  cout << endl;
  bool visited[graph.size()] = {false};
  dfs_rec(graph, visited, 0);
}

File: algorithms/graph/dfs.py
from queue import LifoQueue

def dfs(graph):
  s = LifoQueue()
  s.put(0)
  visited = [False] * len(graph)
  visited[0] = True

  while not s.empty():
    i = s.get()
    print(i)

    for j in reversed(graph[i]):
      if not visited[j]:
        s.put(j)
        visited[j] = True

def dfs_recursive(graph):
  visited = [False] * len(graph)
  _dfs_recursive(graph, visited, 0)

def _dfs_recursive(graph, visited, i):
  if visited[i]: return
  visited[i] = True
  print(i)
  for j in graph[i]: _dfs_recursive(graph, visited, j)

graph = [[1,3],[0,2],[1,3,4],[0,2],[2]]
dfs(graph)
print()
dfs_recursive(graph)
